services/workflow/workflow-manager/src/main/java/com/milesight/beaveriot/rule/manager/service/WorkflowLogService.java

```diff
@@ -0,0 +1,125 @@
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.event.EventListener;
+import org.springframework.data.domain.Page;
+import org.springframework.stereotype.Service;
+import org.springframework.util.StringUtils;
+
+import java.util.List;
+
+@Service
+@Slf4j
+public class WorkflowLogService {
+    @Autowired
+    WorkflowService workflowService;
+
+    @Autowired
+    WorkflowRepository workflowRepository;
+
+    @Autowired
+    WorkflowLogRepository workflowLogRepository;
+
+    @Autowired
+    WorkflowLogDataRepository workflowLogDataRepository;
+
+    public Page<WorkflowLogResponse> searchLogs(Long flowId, SearchWorkflowLogsRequest request) {
+        if (request.getSort().getOrders().isEmpty()) {
+            request.sort(new Sorts().desc(WorkflowPO.Fields.id));
+        }
+
+        return workflowLogRepository
+                .findAll(f -> f
+                        .eq(WorkflowLogPO.Fields.flowId, workflowService.getById(flowId).getId())
+                        .like(StringUtils.hasText(request.getStatus()), WorkflowLogPO.Fields.status, request.getStatus()), request.toPageable()
+                ).map(workflowLogPO -> WorkflowLogResponse.builder()
+                        .id(workflowLogPO.getId().toString())
+                        .startTime(workflowLogPO.getStartTime())
+                        .timeCost(workflowLogPO.getTimeCost())
+                        .status(workflowLogPO.getStatus())
+                        .build()
+                );
+    }
+
+    public WorkflowLogDetailResponse getLogDetail(Long logId) {
+        WorkflowLogPO wl = workflowLogRepository.findById(logId).orElse(null);
+        if (wl == null) {
+            throw ServiceException.with(ErrorCode.DATA_NO_FOUND.getErrorCode(), "Log not found!").build();
+        }
+
+        // check exists
+        workflowService.getById(wl.getFlowId());
+
+        WorkflowLogDataPO wld = workflowLogDataRepository.findById(logId).orElse(null);
+        if (wld == null) {
+            throw ServiceException.with(ErrorCode.DATA_NO_FOUND.getErrorCode(), "Log detail not found!").build();
+        }
+
+        return WorkflowLogDetailResponse.builder()
+                .id(wl.getId().toString())
+                .status(wl.getStatus())
+                .timeCost(wl.getTimeCost())
+                .startTime(wl.getStartTime())
+                .traceInfo(JsonHelper.fromJSON(wld.getData(), new TypeReference<List<NodeTraceInfo>>() {}))
+                .build();
+    }
+
+    public void cleanFlowLogs(List<Long> flowIds) {
+        List<WorkflowLogPO> workflowLogPOList = workflowLogRepository.findAll(f -> f.in(WorkflowLogPO.Fields.flowId, flowIds.toArray()));
+        workflowLogRepository.deleteAllInBatch(workflowLogPOList);
+        workflowLogDataRepository.deleteAllByIdInBatch(workflowLogPOList.stream().map(WorkflowLogPO::getId).toList());
+    }
+
+    @EventListener
```

Comment:
We recommend that you use asynchronous mode to save logs

Reply:
Updated.